### PR TITLE
Improved debian installer

### PR DESCRIPTION
**Pull request by mimooh**

### CODE REVIEW

File: debian/config [hunk 1]
```diff
@@ -2,6 +2,15 @@
 
 set -e
 
+# There are two questions open:
+
+# wartool -m  is inconsistently documented. It can be easily added here, but is it still part of wartool?
+
+# how should the Expansion Pack (Beyond the Dark Portal) be handled? I've
```
Comment (timfel) [inline debian/config]: Sorry, was working down my mailbox, so commented on the closed PR.

Anyway, the expansion CD can be extracted on its own and the game will work, but to get both the ToD and BtDP campaigns you need to extract first from the base game and then from the expansion CD. Alternatively, I'm now building the wargus debian packages with StormLib, so the Battle.net Edition should also work and then you get both campaigns from one CD.

File: debian/config [hunk 2]
```diff
@@ -2,6 +2,15 @@
 
 set -e
 
+# There are two questions open:
+
+# wartool -m  is inconsistently documented. It can be easily added here, but is it still part of wartool?
```
Comment (timfel) [inline debian/config]: -m isn't actually used anymore, it should be removed from the wartool.cpp help text

**Comment by a-detiste** [comment]
..BTW, if you really think "Debconf protocol is a nightmare", you should consider implemting this as a game-data-packager plugin.

There's for example one for lgeneral that call the dedicated `lgc-pg` tool that is similar to `wartool`.

https://github.com/a-detiste/game-data-packager/blob/master/game_data_packager/games/lgeneral.py

The quake2 one will call gcc & make for example.

https://github.com/a-detiste/game-data-packager/blob/master/game_data_packager/games/quake2.py

The datachecksumming & packaging code is not there, because it is in the core engine shared by like 300 packages.
When you have some .deb's it's super easy to launch a lan party, just stuff those in a private repository
(reprepro + small HTTP server) and tell people to `apt-get install warcraft2-data`

**Comment by mimooh** [comment]
> ..BTW, if you really think "Debconf protocol is a nightmare", you
> should consider implemting this as a game-data-packager plugin.
I think I figured it out finally, so now it's fine :) I just couldn't
get the idea from the documentation about how these variables are
passed.

-- 
Karol Kreński

File: debian/config (hunk 1 above)
Comment (mimooh) [inline debian/config]: So we have 3 scenarios:
1. ToD alone
2. BtDP alone
3. ToD + BtDP

I propose to reduce the installer to:
1. Provide one dataset (ToD alone or BtDP alone)
2. Provide two datasets (ToD + BtDP)

Is that ok?

-- 
Karol Kreński

> Anyway, the expansion CD can be extracted on its own and the game will
> work, but to get both the ToD and BtDP campaigns you need to extract
> first from the base game and then from the expansion CD.
> Alternatively, I'm now building the wargus debian packages with
> StormLib, so the Battle.net Edition should also work and then you get
> both campaigns from one CD.

**Comment by timfel** [comment]
I've tried this locally and it looks good to me. Do you think it is ready to merge?

**Comment by timfel** [comment]
Or, while you're at it, do you want to also split the tales data into a separate debian package?

**Comment by a-detiste** [comment]
& please update to debhelper 9 :smile:

**Comment by a-detiste** [comment]
the `-doc` package is failing on me:

```
make[1]: Entering directory '/home/tchet/git/stratagus'
dh_install
cp: impossible d'évaluer « debian/tmp/usr/share/doc »: Aucun fichier ou dossier de ce type
dh_install: cp --reflink=auto -a debian/tmp/usr/share/doc debian/stratagus-doc//usr/share/ returned exit code 1
debian/rules:46: recipe for target 'override_dh_install' failed
```

**Comment by a-detiste** [comment]
`W: wargus: malformed-question-in-templates wargus/extractdata` this one is for you

**Comment by mimooh** [comment]
In case I am the "you" in all these posts :) 

1. Yes, I think the code is ready to merge, I haven't found problems
with it. One thing that is not really solved is how to use two CD's at
the same time, but I'd say it was not better before. So my changes
shouldn't at least worsen anything. If we are fine with:

So we have 3 scenarios: 1. ToD alone 2. BtDP alone 3. ToD + BtDP

then I could tweak the installer a little, perhaps during Christmas.

2. This lpc-pg thing, perhaps a good approach, but for my needs, for LAN
parties I just install on one machine and then copy the whole
/usr/share/games/stratagus to all clients. So I don't have much
motivation for playing with it, sorry :)

3. Packaging tales into another deb seems reasonable. I have never
created a deb package, but it doesn't seem that scary. After all I am
a debconf master now! :)

4. & please update to debhelper 9. No idea what this means, but I may
see about this.

5. What are the full commands for these issues?
a) the -doc package is failing on me
b) W: wargus: malformed-question-in-templates wargus/extractdata

**Comment by a-detiste** [comment]
1. I put my games assets in .deb files, the deb files in a local repository (reprepro), the I just apt-get the games; that scales very well & if there was a bug everyone get the update.
(& they have unattended-upgrades enabled)

2. pleaaase ! because when wartool overwrites "tales..." assets with original Warcraft 2 assets; that makes a lot of security tool complains about corrupted files (`dpkg --verify`, tripwire) and then Warcraft II assets will be replaced again by Tales assets on next update

3. it's about the one-character file in debian/compat & Build-Depends on debhelper (>= 9).
debhelper is the tool that automates 90% of debian/rules that would be extra-long (like it was 15 years ago) without it.
https://wiki.debian.org/IntroDebianPackaging

4. `debuild -us -uc`; that calls `lintian` at the end
   a) I have to cheat in two places to get this compiling.
        1) hide Sqlite so metaserver doesn't get built (this also fails)
        2) create empty /doc & /include

```diff
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -652,7 +652,7 @@ find_package(OpenGL)
 find_package(OpenGLES)
 find_package(SDL_gles)
 
-find_package(Sqlite)
+#find_package(Sqlite)

--- a/debian/rules
+++ b/debian/rules
 override_dh_install:
+       mkdir -p debian/tmp/usr/share/doc
+       mkdir -p debian/tmp/usr/include
        dh_install
```

b) that's the output of lintian; a tool that autodetect errors in packaging

It's a mandatory step to get package in the official archive;
but people also use it on their unofficial archive to detect bugs.

Now, maybe this warning is only relevant to master and not your PR.

```
W: stratagus: hardening-no-relro usr/bin/png2stratagus
I: stratagus: hardening-no-fortify-functions usr/bin/png2stratagus
I: stratagus: spelling-error-in-binary usr/games/stratagus-dbg unknow unknown
I: stratagus: spelling-error-in-binary usr/games/stratagus-dbg weigth weight
W: stratagus: hardening-no-relro usr/games/stratagus-dbg
I: stratagus: hardening-no-fortify-functions usr/games/stratagus-dbg
P: stratagus: no-upstream-changelog
W: stratagus: debian-changelog-line-too-long line 11
I: stratagus: using-first-person-in-description line 14: I
I: stratagus: using-first-person-in-description line 15: I
W: stratagus: package-section-games-but-has-usr-bin
W: stratagus: binary-without-manpage usr/bin/png2stratagus
W: stratagus: binary-without-manpage usr/games/stratagus-dbg
```

**Comment by a-detiste** [comment]
1. as long as wartool remain a generic enough tool, I really don't mind

**Comment by mimooh** [comment]
a-detiste, you seem much more into all this deb building business than me. In fact wargus was my first touch of deb building. I will try to fix at least some of the issues and I promise to at least try to separate tales from warcraft. The idea would be to either install wargus or tales (or both) and the packages will be totally independent. They will have their own directories in /usr/share/games/stratagus/. Is that fine?

**Comment by a-detiste** [comment]
let's say that it's a pain that I'm used to. Maybe wargus is not the easiest target, but it's better to work on something a bit complicated albeit rewarding than something easy you don't care about. Now you're not on your own, there are things likes #debian-mentors or debian-games-devel@ ML with people happy to help new packagers.

The Debian Freecraft package had been orphaned in 2014, so I guess I'll "adopt" it
and create a fake "freecraft" package that pull in wargus + tales without the debconf prompt.

https://packages.qa.debian.org/f/freecraft.html
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=749043

If this PR is done, it could be merged before starting working on something else.

In my PR I was carefull not to stomp over your changes.

As far as I understand, now, Aleona's Tales & Warcraft II share the same directory structures, so one can never have both at same moment; which is unfortunate. So packaging on it's own is not enough to make games co-installable, some changes must happens in core wargus.

The we'll end up with this depency graph:

```
stratagus ----- wargus ----- aleonas-tales-data
                         |
                         \-- warcraft2-data --- warcraft2-music
```

**Comment by mimooh** [comment]
> The we'll end up with this depency graph:
> stratagus ----- wargus ----- aleonas-tales-data
>                          |
>                          \-- warcraft2-data --- warcraft2-music

Is this the resulting directory structure or is it some apt-get dependencies?

If the two datasets are going to be separated it seems like a good
moment to have all the directory paths adjusted. We probably agree that

	/usr/share/games/stratagus/wargus
	/usr/share/games/stratagus/aleonas_tales

would be desired, but perhaps hard to achieve due to what's deep in
wargus core, right? What is aleonas_tales anyway? Is it exactly the same
wargus scripts but only gfx and sounds different from warcraft2? And
perhaps I make a sin now, but does anyone ever play this aleonas_tales?
Is it worth bothering?

-- 
Karol Kreński

**Comment by timfel** [comment]
Separating might not be that hard if we can adjust the cmake install to also create a second installation under `stratagus/tales` and (maybe also from cmake) to rename the wargus binary and pixmap to `aleonas_tales`. Then we strip the install under `stratagus/wargus` by removing the graphics folder. Does that sound reasonable?

**Comment by mimooh** [comment]
> removing the graphics folder. Does that sound reasonable?
Well, I think I get the idea of what you are talking and since I don't
consider myself a wargus hacker at all, I agree on this strategy. I will
start with analyzing cmake and even before I will check what aleona and
warcraft are all about. 

-- 
Karol Kreński

**Comment by a-detiste** [comment]
I never tought it would be really hard to do, if the engine is well made the path shouldn't be encoded everywhere but once in a `config.h` or something; I just wanted some ACK from maintainers so that they understood why this was needed & wether would be ok to merge it.

This is not Debian specific at all, Windows users too wouldn't want the assets from the CD being unexpectedly replaced by Aleonas Tales one's on upgrade.

Aleonas Tales has it's own gfx, sounds but also maps & scenarios.
Some people would prefer this 100% free game over Warcraft II & some would prefer both;
so as it's not that hard we can try to please everyone.


Yup  it's just depedency graph where right parts depends on part linked to the left.

There would be two version of warcraft2-data providing the same files at the same location:
- one .deb almost empty with only the debconf script generated by this git three
- the .deb generated by my packaging tool

Both of these would involve wartool at some point.

> rename the wargus binary and pixmap to aleonas_tales

Or `/usr/games/wargus` can be symlinked as `/usr/games/aleonas_tales` to mean "the two datasets are installed, but this time I prefer Aleonas Tales over Warcraft II (the default)".
This symlink would be shipped by the `aleonas-tales-data` package

**Comment by timfel** [comment]
The paths are not hardcodes anywhere at all, the data

**Comment by timfel** [comment]
Sorry for the close ... fat finger problem... Anyway, the paths are not hardcoded, they are defines from cmake (for unix) and taken from the registry install path (for windows). So it's easy enough to have both datasets.

**Comment by mimooh** [comment]
Am I right that wartool is a conversion tool of warcraft gfx, sounds,
and some warcraft-whatever? In case of Aleona wartool or pudconvert are
not needed, right?

Here is how I understand what is to be done:

	a) wargus.deb:
		rm -rf graphics, sounds and all else related to aleonas
		/usr/share/games/stratagus/wargus
		/usr/games/wargus
	
	
	b) aleonas_tales.deb
		cp -r wargus@github aleonas_tales
		s/wargus/aleonas_tales/g in all files
		try to build, tweak this and that
		skip all the installation dialog as it is not needed for anything
		skip wartool/pudconvert as aleonas is already in the proper format
		/usr/share/games/stratagus/aleonas_tales
		/usr/games/aleonas_tales


If this is right I feel I am caplable of implementing it.

**Comment by a-detiste** [comment]
Hi,

You should be able to generate all the output .deb's in a single run, just replace the `cp` & `rm`with appropriate `mv` 's. The package should remain buildable with a simple `debuild -us -uc -b`

Example, here to build corsix-th & corsix-th-data, I have to move files around 
in `debian/corsix-th` & `debian/corsix-th-data`
http://anonscm.debian.org/cgit/pkg-games/corsix-th.git/tree/debian/control
http://anonscm.debian.org/cgit/pkg-games/corsix-th.git/tree/debian/rules

"s/wargus/aleonas_tales/g in all files"
this should happens once-for -all in the upstream part of the git tree (outside /debian),
bot by weird logic in debian/rules.

"skip wartool/pudconvert as aleonas is already in the proper format"
something like:
```shell
mv wartool debian/wargus/usr/games/
```

"skip all the installation dialog as it is not needed for anything":
that's just 
```shell
cd debian
git mv config warcraft2-data.config
git mv postinst warcraft2-data.postinst
git mv postrm warcraft2-data.postrm
git mv templates warcraft2-data.templates

cat >> control << EOF
Package: warcraft2-data
Architecture: all
Depends: wargus, ${misc:Depends}
Pre-Depends: cdparanoia, ffmpeg2theora
Description: Warcraft II data game set for the Stratagus engine
 Wargus can be used to play Warcraft II from Blizzard Entertainment.
 .
 When this package is installed, a script will prompt you
 to provide the original Warcraft II CD
 and then extract the game data files.
EOF
```

```
c) warcraft2-data: empty package with only the postinst script
```

> I feel I am caplable of implementing it.

Just give it a try :smile:

**Comment by mimooh** [comment]
So I gave it a try, as you suggest, a-detiste :)
https://github.com/mimooh/aleonas-tales

After I am done with aleonas-tales deb, I will start removing stuff from wargus. 
The transformation from wargus to aleonas-tales is fully automatic:
https://github.com/mimooh/aleonas-tales/blob/master/00convert2tales.sh

aleonas-tales can be built and the deb can be installed, but the game cannot be started:
	# aleonas-tales 
[string "/usr/share/games/stratagus/aleonas-tales/scripts/stratagus.lua"]:37: '=' expected near '-'

I will have a look at it later.

**Pull request merged**